File: packages/platform/src/platform.ts
import { spawnSync } from 'node:child_process'

export function assertMacOS(): void {
  if (process.platform !== 'darwin') {
    const msg = 'whirlpool only supports macOS in v1.'
    const color =
      process.stderr.isTTY &&
      process.env.NO_COLOR == null &&
      process.env.FORCE_COLOR !== '0'
    if (color) {
      process.stderr.write(`\x1b[31m${msg}\x1b[0m\n`)
    } else {
      process.stderr.write(`${msg}\n`)
    }
    process.exit(1)
  }
}

/**
 * Opens System Settings / Preferences near **Full Disk Access** via `open` and
 * Apple’s `x-apple.systempreferences:` scheme so the user can grant access to
 * the host app (Terminal, Cursor, iTerm, etc.).
 */
export function openFullDiskAccessSettings(): void {
  if (process.platform !== 'darwin') return
  spawnSync(
    '/usr/bin/open',
    ['x-apple.systempreferences:com.apple.settings.PrivacySecurity.extension'],
    { stdio: 'ignore' },
  )
}

File: whirlpool/utils/platform.py
import subprocess
from sys import platform

OS_NAMES = {
    "darwin": "macOS",
    "linux": "Linux",
    "win32": "Windows",
}


class Platform:
    @staticmethod
    def assert_macos():
        os_name = OS_NAMES.get(platform)
        if os_name is None:
            raise RuntimeError(f"Unknown OS: {platform}")

        if platform != "darwin":
            raise RuntimeError(f"This script must be run on MacOS, not {os_name}")

    @staticmethod
    def get_fulldisk_permissions():
        if platform != "darwin":
            return False

        try:
            result = subprocess.run(
                [
                    "/usr/bin/open",
                    "x-apple.systempreferences:com.apple.settings.PrivacySecurity.extension",
                ],
                capture_output=True,
                text=True,
            )

            return "system.preferences" in result.stdout
        except subprocess.CalledProcessError:
            return False
